site/content/features/recordpatterns.md

```diff
@@ -0,0 +1,619 @@
+
+```
+if (obj instanceof UnixDomainPrincipal p) {
+   var u = p.user();
+   var g = p.group();
+   // Do something with u and g
+}
+```
+
+Instead of `var`, you can also use the actual types of the components:
+
+```
+if (obj instanceof UnixDomainPrincipal(UserPrincipal u, DomainPrincipal g)) {
+   // Do something with u and g
+}
+```
+
+Either way, the syntax is meant to remind you of variable declarations. 
+
+You can also use a record pattern in `switch`:
+
+```
+switch (obj) {
+   case UnixDomainPrincipal(var u, var g):
+      // Do something with u and g
+      break;
+   default:
+      break;
+}
+```
+
+That's potentially nice, but how often does it happen that you have an `Object` that might be a record instance? To see more interesting examples, we need multiple record classes. It gets even better when they extend a sealed interface because then the `switch` can test for exhaustiveness.
+
+## A Sealed Record Family
+
+Ever since Java 1.4, there has been a `CharSequence` interface with methods
+
+```
+char charAt(int index);
+int length();
+CharSequence subSequence(int start, int end);
+String toString();
+```
+
+Java 8 added a couple of default methods, and Java 11 a static method. We ignore them for this example.
+
+The interface is implemented by `StringBuilder`, the legacy `StringBuffer`, `java.nio.CharBuffer`, a Swing class, and of course `String`. It is mostly used to write code that works with both `String` and `StringBuilder`.
+
+We want to manipulate subsequences. They could touch the beginning or the end, or they lie in the middle. This is where we get a sealed interface and three records:
+
+```
+sealed interface SubSequence extends CharSequence permits Initial, Final, Middle { /* ... */ }
+
+record Initial(CharSequence seq, int end) implements SubSequence { /* ... */ }
+record Final(CharSequence seq, int start) implements SubSequence { /* ... */ }
+record Middle(CharSequence seq, int start, int end) implements SubSequence { /* ... */ }
+```
+
+Of course, we need to implement the `CharSequence` methods. That's easily done in the superinterface with a pattern match:
+
+```
+default int length() {
+   return switch (this) {
+      case Initial(var __, var end) -> end;
+      case Final(var seq, var start) -> seq.length() - start();
+      case Middle(var __, var start, var end) -> end - start;
+   };
+}
+```
+
+No `default` is required because we provided cases for all classes that implement the sealed interface.
+
+Note the double underscore for the variables that we don't care about. (A single underscore is a Java keyword, held in reserve for future use.)
+
+The following sandbox contains the complete example. Note that a record pattern can have a guard:
+
+```
+case Initial(var seq, var end) when s == 0
+```
+
+{{< sandbox version=java19 preview="true" mainclass="SubSequence" >}}{{< sandboxsource "SubSequence.java" >}}
```

Comment:
To actually execute the Main class you probably need to specify

* `mainclass="Main"`
* `sandboxsource "Main.java"`

In the current setup I only see the compiler warnings. And I wonder why I don't see an error that SubSequence has no main method. Something I need to look after in the sandbox.

Reply:
Forgot to update the regex in the Markdown converter. Fixed.